### ISSUE TITLE
LSP syntax highlighting

### ISSUE THREAD

**Issue by kirawi** [helix-editor/helix#814]
This will likely require us to revise how we handle syntax highlighting as a whole to acclimate to methods other than tree-sitter.

**Comment by archseer** [comment]
I'm not a big fan of the LSP highlighting spec. It requires a lot of back and forth traffic with the LSP to do highlighting in real time. VSCode should adapt tree-sitter already..

**Comment by kirawi** [comment]
I plan to make it optional in `config.toml`, but there are some cases in which tree-sitter can't be used easily.

**Comment by JakeHL** [comment]
> I'm not a big fan of the LSP highlighting spec. It requires a lot of back and forth traffic with the LSP to do highlighting in real time. VSCode should adapt tree-sitter already..

Does this mean this won't ever be a first class feature of helix? I'm not really fussed either way. I only ask because I'm working on a tool to convert themes between several text editors formats and this would factor into the implementation longer term.

**Comment by LouisGariepy** [comment]
While LSP is heavy, tree-sitter does not provide comparable highlighting. One solution could be to have LSP as an opt-in, with tree-sitter as the default.

As petty as it sounds, the lack of syntax highlighting is one of the very few things that hold me from daily driving Helix. Being able to jump in and see the same theming quality as I'm used to is a must for me, and I imagine other prospective users as well.

**Comment by pickfire** [comment]
@LouisGariepy IIRC in the release notes I remember there is `space-h` select the same symbols in the function, that way it can be similar like part of the syntax highlight feature.

**Comment by LouisGariepy** [comment]
Could a maintainer point out the conditions for an acceptable solution then? If you simply don't want this feature in Helix (which is totally fair!), maybe you could close or mark as "wontfix" to avoid confusion.

**Comment by JakeHL** [comment]
@LouisGariepy 
> Being able to jump in and see the same theming quality as I'm used to is a must for me, and I imagine other prospective users as well.

I'd be interested to see a side by side comparison with the theme/languages you use if you don't mind? In my experience, it's the opposite. I quite often find tree sitter highlighting being better than the alternatives.

**Comment by univerz** [comment]
> I'd be interested to see a side by side comparison with the theme/languages you use if you don't mind?

i miss mutable variables (in rust) with `bold` modifier.

**Comment by archseer** [comment]
That's doable with a custom `locals.scm` query.

**Comment by poliorcetics** [comment]
> That's doable with a custom `locals.scm` query.

Is it though ? They're talking about highlighting the variable as mutable each time it's used, not just at the definition point.

I honestly don't know if this is possible, if it is it seems like a quick win to add it to helix

**Comment by kirawi** [comment]
Yeah, that's doable with a `locals.scm`. It will track the use of an identifier throughout the scope and apply the color it was highlighted with at the definition.

**Comment by LouisGariepy** [comment]
Alright! I jotted down some basic code to see the potential differences. I tried to emphasize differences due to "limitations", not stylistic ones. There might be more that I'm missing for complex code bases, this is just basic stuff.

Here goes: 

#### Public items are distinct from private ones.
Note that this holds for any item that can be made public, not just functions.

![public_item](https://user-images.githubusercontent.com/68077556/188671692-3c2d0fcb-c1d8-416e-9472-5ead7a514b41.png)

![public_item_helix](https://user-images.githubusercontent.com/68077556/188671698-d9fefdca-d54d-4691-84cf-c8fd674d75d8.png)

#### Mutable variables and operators are distinct from immutable ones.
Mutable functions/method calls are likewise distinct (not shown in image).

![mutable_variables](https://user-images.githubusercontent.com/68077556/188671829-cdaf1baa-e559-48e7-91c5-23a7a3332c27.png)

![mutable_variables_helix](https://user-images.githubusercontent.com/68077556/188671844-129e8cbd-90f5-4491-a8fc-1f7a4386d455.png)

#### Variables in format strings are recognized as such (distinct from the rest of the string).

![format_variables](https://user-images.githubusercontent.com/68077556/188671234-c4ccb5b0-9035-4236-955e-dcacfaf9ddba.png)

![format_variables_helix](https://user-images.githubusercontent.com/68077556/188671305-814cb69e-bc0d-453d-99ab-78ce265aa258.png)

#### Doc comments are distinct from "regular" comments.
Might be hard to see (stylistic choice), but the doc comments are lighter in color.

![doc_comments](https://user-images.githubusercontent.com/68077556/188671903-080d084e-bd14-4627-b380-0a1bb00baa8b.png)

![doc_comments_helix](https://user-images.githubusercontent.com/68077556/188671945-db9cf525-93dd-4153-8b35-4448095da84b.png)

#### Generic types are distinct from their trait bounds.

![trait_bounds](https://user-images.githubusercontent.com/68077556/188671981-5cdfc4a5-d180-46b4-b106-c280285ce926.png)

![trait_bounds_helix](https://user-images.githubusercontent.com/68077556/188672005-a1e6832c-2e3a-4d17-a2d8-38ba864006b7.png)

#### Macro rules binding arguments are distinct from constant patterns.
`$e` and `$es` are distinct in color.

![macro_rules](https://user-images.githubusercontent.com/68077556/188672195-8f10d94a-11ce-4519-b733-8440ef9fa91e.png)

![macro_rules_helix](https://user-images.githubusercontent.com/68077556/188672204-59b9faf7-acc8-4d31-b7b3-8c933a879c4d.png)

#### Attribute syntax is recognized properly.

![attributes](https://user-images.githubusercontent.com/68077556/188672251-959750a3-0b2b-4407-a781-8d7fddab14a5.png)

![attributes_helix](https://user-images.githubusercontent.com/68077556/188672260-4972d107-1920-4250-8f01-34abeafbbf71.png)


#### Imports (`use`s) are properly highlighted according to their types.

![imports](https://user-images.githubusercontent.com/68077556/188672370-9b223412-04f5-4a76-b40f-2f9723d07c6c.png)

![imports_helix](https://user-images.githubusercontent.com/68077556/188672378-49c688f1-fec3-4bef-9d05-08fa71535f12.png)

#### Trait methods are distinct from non-trait methods.

![trait_methods](https://user-images.githubusercontent.com/68077556/188672498-5cedab0d-79ef-4a03-81a3-d2240b429534.png)

![trait_methods_helix](https://user-images.githubusercontent.com/68077556/188672535-e9384728-7231-4172-88cc-0c359222aa51.png)

#### Enum variants are distinct from structs and function calls

![enum_variants](https://user-images.githubusercontent.com/68077556/188673177-1cfb0142-3a6f-48ef-8145-c22a90d4530c.png)

![enum_variants_helix](https://user-images.githubusercontent.com/68077556/188673201-2f4533ca-4b88-4415-917a-216abc630109.png)

# Conclusion
While some of these features might be possible to implement in tree-sitter, I'd like to encourage you to think of the practicality of doing so. I have put many hours into trying to make my "perfect" tree-sitter theme, and I had to give up because I simply could not find a decent/possible way to make some features work.

For end users, this results in two things:
1) Being unable to use the plethora of LSP themes that already exist.
2) The themes that do exist in tree-sitter form do not have as extensive highlighting capabilities.

**Comment by kirawi** [comment]
Most of these are currently achievable by modifying the queries except for imports and accurate type highlighting. With https://github.com/helix-editor/helix/issues/1252 however, I think that would be addressed as well.

There are certainly limitations to tree-sitter, such as not being able to highlight markdown in doc comments, but maybe that's something that could be worked on in the future.

**Comment by LouisGariepy** [comment]
@kirawi Are aware of tree-sitter themes with these features, or are you speculating that it *could* be done? 

If there's a simple way to make most of these work already with tree-sitter, I'll gladly switch over :smile:.

**Comment by kirawi** [comment]
It looks like you deleted your last comment, but I think you have a misunderstanding of how tree-sitter works. From how I understand it, tree-sitter a parser generator with an AST geared towards syntax highlighting. However, you also need to write queries to capture AST nodes such that they can be highlighting in themes.

You can check [the tree-sitter playground](https://tree-sitter.github.io/tree-sitter/playground) to see some of what these ASTs look like.

See:
https://tree-sitter.github.io/tree-sitter/syntax-highlighting#queries
https://tree-sitter.github.io/tree-sitter/using-parsers#pattern-matching-with-queries

**Comment by JakeHL** [comment]
@LouisGariepy Thanks for that rundown. I think you've demonstrated why it would be a useful feature.

I do think that the main thing it gives us that tree sitter doesn't is type related highlighting. I'm pretty sure with enough time you could create tree-sitter queries for anything that is purely syntax focused.

Here's an example of doc comments & string interpolation already working in Typescript in helix. I might be using a modified Everforest theme, I can't remember.

![Screen Shot 2022-09-07 at 11 10 59](https://user-images.githubusercontent.com/9119973/188853272-2145a5a0-93a4-419c-a64e-0c93bb3b29a0.png)

**Comment by archseer** [comment]
Yeah, tree-sitter grammars are parsers that build an in-memory AST of the file. Then the highlighting queries are like selectors that tag specific parts of the tree with span names, and themes can hook into these spans to highlight them as they wish.

---

> Public items are distinct from private ones.

This is as simple as adding a highlighting query that themes can hook into.

> Mutable variables and operators are distinct from immutable ones.

Can be done with a locals query.

> Variables in format strings are recognized as such (distinct from the rest of the string).

Probably needs a change in the parser, but it could be injected into such strings and the embedded parts interpolated. It works in other languages.

> Doc comments are distinct from "regular" comments.

This is something I've been waiting for https://github.com/tree-sitter/tree-sitter-rust/pull/128

> Generic types are distinct from their trait bounds.

This is as simple as adding a highlighting query that themes can hook into.

> Macro rules binding arguments are distinct from constant patterns.

Not sure yet since macros produce a token tree but probably possible.

> Attribute syntax is recognized properly.

Fixed in https://github.com/helix-editor/helix/commit/301f5d7cf704c2f2e4fc53e258125e39d1845176

> Imports (uses) are properly highlighted according to their types.
> Trait methods are distinct from non-trait methods.

These two are to do so yeah, it could be LSP specific.

> Enum variants are distinct from structs and function calls

Probably possible with locals.

---

I wish LSP provided these highlights as an augmented layer on top of regular highlighting instead of completely overriding highlighting and churning out tons and tons of spans per keypress.

Anyway, for this to be implemented, a vec of produced spans would have to sit somewhere on the document, that would then convert to a `HighlightIter` and hook into our regular rendering mechanism without changes.

**Comment by archseer** [comment]
> I do think that the main thing it gives us that tree sitter doesn't is type related highlighting.

I think in combination with [stack graphs](https://github.blog/2021-12-09-introducing-stack-graphs/) we wouldn't just get go to definition support but also a way to determine the item type since we can peek at the definition

**Comment by kirawi** [comment]
A discussion on how LSP syntax highlighting can be gracefully implemented was discussed on Matrix:

Tree-sitter will be the default highlighter, while range-based LSP syntax highlighting will be applied asynchronously. Additionally, we would compute the correct highlights for locals based on what the LSP syntax highlighter sends.
